### PR TITLE
fix: convert telemetry ulids to uuids

### PR DESCRIPTION
**Pull request by colesnodgrass**
## What
- fix airbytehq/airbyte-internal-issues#7976
- convert the user and session ulid identifiers to uuid identifiers
- attempts to generate a uuid via the following code
   ```bash
   TelemetryUUID() {
    local uuid

    if command -v uuidgen > /dev/null; then
      uuid=$(uuidgen)
    elif [ -f /proc/sys/kernel/random/uuid ]; then
      uuid=$(cat /proc/sys/kernel/random/uuid)
    elif command -v openssl > /dev/null; then
      local opensslUUID
      opensslUUID=$(openssl rand -hex 16)
      uuid=${opensslUUID:0:8}-${opensslUUID:8:4}-${opensslUUID:12:4}-${opensslUUID:16:4}-${opensslUUID:20:12}
    elif command -v python > /dev/null; then
      uuid=$(python -c "import uuid; print(uuid.uuid4())")
    else
      uuid=""
    fi

    echo $uuid
   }
   ```
- write the data uuid data out as `anonymous_user_uuid` in the `~/.airbyte/analytics.yml` file
    - will also write out the prior ulid value as `anonymous_user_id` for _temporary_ backwards compatibility  

## Can this PR be safely reverted and rolled back?
- [x] YES 💚
- [ ] NO ❌

### CODE REVIEW

**Review by aaronsteers — CHANGES_REQUESTED** [review]
A few meta-points I would like to address:

1. Maybe I'm missing context, but I think a clean cut to the new UUID is preferable to creating a new field. Any reason not to just replace the existing ULID with a UUID whenever a ULID is observed?
2. Apologies if I missed it, but I don't see a map between old values and new values in this implementation - which is important to preserve the original identity and (more importantly) to preserve distinct user counts. I'm aware of two ways to preserve lineage: 
   1. Emit an "id changed" event with old and new identifiers.
      - This may be a bit more work but it means we can join from old ID to new ID for deduping purposes in the analytics warehouse.
   2. Derive the new UUID deterministically from the old ULID.
       - I think this option is my personal preference, as it is "stateless" and doesn't require any mapping tables.

Curious to your thoughts on these. I'm also going to reread the PR changes in case any of the points of uncertainty I called out have answers built-in that I missed in the first read through.

**Comment by colesnodgrass** [comment]
@aaronsteers 
> 1. Maybe I'm missing context, but I think a clean cut to the new UUID is preferable to creating a new field. Any reason not to just replace the existing ULID with a UUID whenever a ULID is observed?

This is discussed in the [parent ticket](https://github.com/airbytehq/airbyte-internal-issues/issues/7758), but it basically boils down to this is a breaking change that doesn't play nice with [abctl](https://github.com/airbytehq/abctl/blob/72a2c268c56f64455866710ab20a24d941746e38/internal/telemetry/config.go#L19-L63) as it requires this field to currently be a `ulid`.  Which means, during the transition from the ulid to uuid, we will need to support both to ensure that all the tools continue to work as expected.

> Emit an "id changed" event with old and new identifiers.
This may be a bit more work but it means we can join from old ID to new ID for deduping purposes in the analytics warehouse.

I _think_ this is possible, if are in the situation where a `ulid` value already exists and new `uuid` value is generated. 

> Derive the new UUID deterministically from the old ULID.
I think this option is my personal preference, as it is "stateless" and doesn't require any mapping tables.

This would be difficult, as we're limited to what a bash script can do.

**Comment by aaronsteers** [comment]
> > 1. Maybe I'm missing context, but I think a clean cut to the new UUID is preferable to creating a new field. Any reason not to just replace the existing ULID with a UUID whenever a ULID is observed?
> 
> This is discussed in the [parent ticket](https://github.com/airbytehq/airbyte-internal-issues/issues/7758), but it basically boils down to this is a breaking change that doesn't play nice with [abctl](https://github.com/airbytehq/abctl/blob/72a2c268c56f64455866710ab20a24d941746e38/internal/telemetry/config.go#L19-L63) as it requires this field to currently be a `ulid`. Which means, during the transition from the ulid to uuid, we will need to support both to ensure that all the tools continue to work as expected.

To be clear, PyAirbyte only cares that it is a string. I'd much prefer a clean cutover than having multiple IDs. If we create a different field here, then `abctl` would report using that ID, while PyAirbyte would continue reporting the other - somewhat defeating the purpose and breaking user counts and new user conts.

Any reason on the `abctl` side not to replace the ID with the UUID one - your script could check the length and/or regex, and if its a ULID (or _not_ a UUID), we can trigger the behavior to simply replace the existing fields' value.

I think we still need a solution for `2.a.` or `2.b.` or similar - either sending an event to bridge the two IDs (any message that contains both will do), or converting that existing ULID to a UUID-shaped value.

File: run-ab-platform.sh
Comment (perangel) [inline run-ab-platform.sh:129]: I was just reading the documentation we currently have on our telemetry and it mentions how to control (as in enable/disable) for other deployment methods but not `abctl`. Perhaps worth throwing something on that page about it

**Comment by colesnodgrass** [comment]
@aaronsteers 
> I think we still need a solution for 2.a. or 2.b. or similar - either sending an event to bridge the two IDs (any message that contains both will do), or converting that existing ULID to a UUID-shaped value.

We don't need this level of tracking for the `run-ab-platform` and `abctl` scripts, so I'm not going to add it here.  If you need this on the `PyAirbyte` side, you should be able to add this without any problems as both `run-ab-platform` and `abctl` will keep the old (`*_id`) value and the new (`*_uuid`) value around.

**Comment by aaronsteers** [comment]
@colesnodgrass  - I still can't approve this PR because all of the issues I noted in my original review still stand.

If this is time sensitive, and it must merge, you can dismiss my prior review as needed for expediency. I don't want to block any important function.

Happy to jump on a call to discuss.

File: run-ab-platform.sh
Comment (colesnodgrass) [reply run-ab-platform.sh:129]: 👍🏻 This should be listed on that page now as well.  I'll ping our docs people about this.

File: run-ab-platform.sh
Comment (aaronsteers) [inline run-ab-platform.sh:70]: Discussed on Zoom. Renaming this to `analyticsID` (probably spelled `analytics_id` in the Yaml) will resolve my concern about conflation of the two fields, and we can still over time migrate specific applications from one to the other.

```suggestion
telemetryUserULID=""
analyticsID=""
```

**Review by aaronsteers — APPROVED** [review]
@colesnodgrass - Thanks for the zoom chat. As discussed, I'm approving with just the one condition of a rename of the new field. It's not ideal to have two fields, but I get the limitations due to strongly-typed reads on this field in `abctl` and why we kind of have to make that call to create a new field.

As discussed, I'll also update PyAirbyte so that it does not fully overwrite the file on cases where the file exists but `anonymous_user_id` does not exist as a key.

This path (with two different-enough-sounding names) feels decently sustainable over the long-run, and is minimally painful if both fields end up living on for longer than we'd like. So, I think that's a good win - and then everything else is basically unblocked to move forward. 🙌 

Cheers. 🚀

**Pull request merged**
